### PR TITLE
Handle 408 responses by retrying the request

### PR DESCRIPTION
**Pull request by dave-r12**
First draft for #1693. In my testing, both Chrome and Firefox repeat the request. Wasn't sure on best approach for the cache tests.

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/http/HttpEngine.java [hunk 1]
```diff
@@ -940,6 +941,17 @@ public Request followUpRequest() throws IOException {
...
+      case HTTP_CLIENT_TIMEOUT:
+        // 408's are rare in practice, but some servers like HAProxy use this response code. The
+        // spec says that we may repeat the request without modifications. Modern browsers also
+        // repeat the request (even non-idempotent ones.)
+        boolean retryableBody = requestBodyOut == null || requestBodyOut instanceof RetryableSink;
+        if (callerWritesRequestBody && !retryableBody) {
+          return null;
+        }
+
+        return userRequest.newBuilder().build();
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http/HttpEngine.java]: Is a copy needed here? You can't just return the original `userRequest` directly?

Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/internal/http/HttpEngine.java]: Gotcha, will change

File: okhttp-tests/src/test/java/okhttp3/CallTest.java [hunk 2]
```diff
@@ -1400,6 +1400,41 @@ private void postBodyRetransmittedAfterAuthorizationFail(String body) throws Exc
...
+  @Test public void getClientRequestTimeout() throws Exception {
+    enqueueRequestTimeoutResponses();
+
+    Response response = client.newCall(new Request.Builder()
+        .url(server.url("/")).build()).execute();
+
+    assertEquals("Body", response.body().string());
+  }
+
+  private void enqueueRequestTimeoutResponses() {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/CallTest.java]: Mostly off topic . . . though it causes duplication, I’m generally a fan of just inlining methods like these. For production code DRY is good. But for tests, I prefer to keep tests self-contained over keeping ’em not-repetitive.

That said, it's matters of degree. Lots of our test suite is built with supporting classes - even MockWebServer started out as a humble test utility here.

File: okhttp-tests/src/test/java/okhttp3/URLConnectionTest.java [hunk 3]
```diff
@@ -2327,6 +2327,53 @@ private void testRedirect(boolean temporary, String method) throws Exception {
...
+    assertEquals("Body", readAscii(connection.getInputStream(), Integer.MAX_VALUE));
+
+    RecordedRequest request1 = server.takeRequest();
+    assertEquals("Hello", request1.getBody().readUtf8());
+
+    RecordedRequest request2 = server.takeRequest();
+    assertEquals("Hello", request2.getBody().readUtf8());
+  }
+
+  @Test public void streamedBodyWithClientRequestTimeout() throws Exception {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/URLConnectionTest.java]: Nice.

**Comment by swankjesse** [comment]
This is awesome.

File: okhttp-tests/src/test/java/okhttp3/CallTest.java (hunk 2 above)
Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/CallTest.java]: Will do, I'll keep that in mind going forward

**Pull request merged**